bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/JsonGenerator.java

```diff
@@ -307,12 +308,22 @@ public void serialize(BValue json) throws IOException {
                 break;
             case TypeTags.MAP_TAG:
             case TypeTags.JSON_TAG:
-                this.startObject();
-                for (Entry<String, BValue> entry : ((BMap<String, BValue>) json).getMap().entrySet()) {
-                    this.writeFieldName(entry.getKey());
-                    serialize(entry.getValue());
+            case TypeTags.RECORD_TYPE_TAG:
+                if (json instanceof BRefValueArray) {
```

Comment:
we may not need this check. any value with JSON will always be a BMap. JSAON arrays will have the array tag. 
I thin there was an issue with stamping earlier, which could be why you were getting an array with a json tag. It is fixed now.